src/registry/AvatarRegistry.sol

```diff
@@ -0,0 +1,541 @@
+
+        if (upKeepId == 0) {
+            revert AvatarNotRegisteredYet(avatar);
+        }
+        if (!underFunded) {
+            revert NotUnderFundedUpkeep(upKeepId);
+        }
+
+        uint96 topupAmount = minUpKeepBal * uint96(ROUNDS_TOP_UP);
+
+        if (LINK.balanceOf(address(this)) < topupAmount) {
+            _swapEthForLink(topupAmount);
+        }
+
+        CL_REGISTRY.addFunds(upKeepId, topupAmount);
+    }
+
+    /// @dev executes the swap from ETH to LINK, for the amount of link required
+    /// @param linkRequired amount of link required for handling the `performUpKeep` task
+    function _swapEthForLink(uint256 linkRequired) internal {
+        uint256 maxEth = (getLinkAmountInEth(linkRequired) / MAX_BPS) * BUMP_UP_BPS;
+        uint256 ethSpent = UNIV3_ROUTER.exactOutputSingle{value: maxEth}(
+            IUniswapRouterV3.ExactOutputSingleParams({
+                tokenIn: WETH,
+                tokenOut: address(LINK),
+                fee: uint24(3000),
+                recipient: address(this),
+                deadline: type(uint256).max,
+                amountOut: linkRequired,
+                amountInMaximum: maxEth,
+                sqrtPriceLimitX96: 0 // Inactive param
+            })
+        );
+        UNIV3_ROUTER.refundETH();
+        emit EthSwappedForLink(ethSpent, linkRequired, block.timestamp);
+    }
+
+    /// @dev carries registration of target contract in CL
+    /// @param targetAddress contract which will be register
+    /// @param gasLimit amount of gas to provide the target contract when
+    /// performing upkeep
+    /// @param name detailed name for the upkeep job
+    /// @return upkeepID id of cl job
+    function _registerUpKeep(address targetAddress, uint256 gasLimit, string memory name)
+        internal
+        returns (uint256 upkeepID)
+    {
+        /// @dev we ensure we top-up enough LINK for couple of test-runs (20) and sanity checks
+        uint256 linkAmount = _getLinkAmount(gasLimit) * ROUNDS_TOP_UP;
+        if (linkAmount < MIN_FUNDING_UPKEEP) {
+            revert NotMinLinkFundedUpKeep();
+        }
+        if (LINK.balanceOf(address(this)) < linkAmount) {
+            _swapEthForLink(linkAmount);
+        }
+
+        /// @dev check registry state before registering
+        (IKeeperRegistry.State memory state,,) = CL_REGISTRY.getState();
+        uint256 oldNonce = state.nonce;
+
+        bytes memory data = abi.encodeCall(
+            IKeeperRegistrar.register,
+            (
+                name,
+                bytes(""),
+                targetAddress,
+                uint32(gasLimit),
+                address(this),
+                bytes(""),
+                uint96(linkAmount),
+                0,
+                address(this)
+            )
+        );
+
+        LINK.transferAndCall(KEEPER_REGISTRAR, linkAmount, data);
+
+        (state,,) = CL_REGISTRY.getState();
+        uint256 newNonce = state.nonce;
+
+        if (newNonce == oldNonce + 1) {
```

Comment:
This returns a `boolean`in success on the call and perhaps we could check if returns true to ensure that the `contractFallback` indeed went fine. but instead we will need to verify that it occurred an increase of nonce in the registry to ensure a new upkeep was automatically registered

Decided just to check that, since it was not increased, means was not properly registered, which may have similar reverting effect as checking the boolean imo

Reply:
here is how i do it in a brownie test:
```
def test_register_new_upkeep(owner, kk, keeper_registrar):
    approved_count = keeper_registrar.getRegistrationConfig()[2]

    kk.register_member(
        "0x697A8DebFbb6bE97c821b482AECc30477F40A102",
        "NewUpkeep",
        1_000_000,
        {"from": owner, "value": 10e18},
    )

    assert keeper_registrar.getRegistrationConfig()[2] > approved_count
```

Reply:
but yea i guess gas wise, checking the return of `transferAndCall` is the most efficient